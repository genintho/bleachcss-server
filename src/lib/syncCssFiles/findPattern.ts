
export default function findPattern(logger, url: string): {name: string, pattern: string} {
    const noScheme = url.replace(/.*:\/\//, '');
    const host = noScheme.split('/', 1)[0];
    const path = noScheme.replace(host, '').split('?', 1)[0];
    let pathName = path;
    let pathPattern = regexEscape(path);
    const query_pieces = (noScheme.replace(host, '').replace(path, '').replace(/^\?/, '')).split('&');
    let queryNamePieces: string[] = [];
    let queryPatternPieces: string[] = [];

    Object.keys(pathVariations).forEach(variation => {
        const {detect, replace} = pathVariations[variation];
        pathName = pathName.replace(detect, variation);
        pathPattern = pathPattern.replace(detect, replace);
    });

    query_pieces.forEach(piece => {
        Object.keys(pathVariations).forEach(variation => {
            const {detect, replace} = pathVariations[variation];
            pathName = pathName.replace(detect, variation);
            pathPattern = pathPattern.replace(detect, replace);
        });
        for (let variation in queryVariations) {
            const {detect, replace} = queryVariations[variation];
            if (detect.test(piece)) {
                queryNamePieces.push(variation);
                queryPatternPieces.push(replace);
                return;
            }
        }
        queryNamePieces.push(piece);
        queryPatternPieces.push(regexEscape(piece));
    });

    let queryName = queryNamePieces.join('&');
    queryName = queryName ? '?' + queryName : queryName;

    let queryPattern = queryPatternPieces.join('&');
    queryPattern = queryPattern ? '\\?' + queryPattern : queryPattern;

    let hostPattern = regexEscape(host.replace(domainDetectRegex, '')) + domainReplaceRegex;
    let hostName = host.replace(domainDetectRegex, '.<domain>');
    return {
        name: hostName + pathName + queryName,
        pattern: hostPattern + pathPattern + queryPattern,
    }
}
function regexEscape(string: string): string{
    return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

const commonSecondLevelDomains = ['ac', 'co', 'com', 'edu', 'gov', 'me', 'net', 'org'];
const domainReplaceRegex = '(\.(' + commonSecondLevelDomains.join('|') + '))?\.[^./]+';
const domainDetectRegex = new RegExp(domainReplaceRegex + '$', 'i');

const pathVariations = {
    '<hash64>': {
        detect: /\b[0-9a-f]{64}\b/gi,
        replace: '[0-9a-f]{64}'
    },
    '<hash40>': {
        detect: /\b[0-9a-f]{40}\b/gi,
        replace: '[0-9a-f]{40}'
    },
    '<hash32>': {
        detect: /\b[0-9a-f]{32}\b/gi,
        replace: '[0-9a-f]{32}'
    },
    '_<hash32>_': {
        detect: /_[0-9a-f]{32}_/gi,
        replace: '_[0-9a-f]{32}_'
    },
    '<hash12>': {
        detect: /\b[0-9a-f]{12}\b/gi,
        replace: '[0-9a-f]{12}'
    },
    '<hash>': {
        detect: /\b[0-9a-f]{10,}\b/gi,
        replace: '[0-9a-f]{10,}'
    },
    '<unixtime>': {
        detect: /\b1[0-9]{9}\b/g,
        replace: '1[0-9]{9}'
    },
    '<unixtimeMillis>': {
        detect: /\b1[0-9]{12}\b/g,
        replace: '1[0-9]{12}'
    },
};
const queryVariations = {
    'timestamp=<timestamp>': {
        detect: /^timestamp=[\d]+$/i,
        replace: 'timestamp=[\\d]+',
    },
    'v=<version>': {
        detect: /^v=([\d.]|[a-f\d])+$/i,
        replace: 'v=([\\d.]|[a-f\\d])+',
    },
    'ver=<version>': {
        detect: /^ver=([\d.]|[a-f\d])+$/i,
        replace: 'ver=([\\d.]|[a-f\\d])+',
    },
    'version=<version>': {
        detect: /^version=([\d.]|[a-f\d])+$/i,
        replace: 'version=([\\d.]|[a-f\\d])+',
    },
};
