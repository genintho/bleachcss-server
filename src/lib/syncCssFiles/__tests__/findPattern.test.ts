import findPattern from '../findPattern';
const testData = require('./patternUrls').default;
import loggerFactory from '../../logger';
const logger = loggerFactory('test');

 test('findPattern should work', () => {
     expect(testData.map((url) => {
         return findPattern(logger, url);
     })).toMatchSnapshot();
});
