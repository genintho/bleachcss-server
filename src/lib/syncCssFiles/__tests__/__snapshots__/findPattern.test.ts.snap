// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findPattern should work 1`] = `
Array [
  Object {
    "name": "sumome-140a.kxcdn.<domain>/virtual/<hash40>/client/css/sme-popup.css",
    "pattern": "sumome-140a\\\\.kxcdn(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/virtual/[0-9a-f]{40}/client/css/sme-popup\\\\.css",
  },
  Object {
    "name": "ucarecdn.<domain>/assets/frontend/public/fonts.<hash32>.css",
    "pattern": "ucarecdn(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/assets/frontend/public/fonts\\\\.[0-9a-f]{32}\\\\.css",
  },
  Object {
    "name": "s.huffpost.<domain>/assets/css.php?f=_quickread.css&v=<version>",
    "pattern": "s\\\\.huffpost(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/assets/css\\\\.php\\\\?f=_quickread\\\\.css&v=([\\\\d.]|[a-f\\\\d])+",
  },
  Object {
    "name": "cdn.wedding-spot.<domain>/css/bootstrap3/bootstrap.min.<hash12>.css",
    "pattern": "cdn\\\\.wedding-spot(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/css/bootstrap3/bootstrap\\\\.min\\\\.[0-9a-f]{12}\\\\.css",
  },
  Object {
    "name": "cdn.wedding-spot.<domain>/css/bootstrap3/bootstrap.min.<hash>.css",
    "pattern": "cdn\\\\.wedding-spot(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/css/bootstrap3/bootstrap\\\\.min\\\\.[0-9a-f]{10,}\\\\.css",
  },
  Object {
    "name": "cdn-asset-mel-1.airsquare.<domain>/www/site.css?timestamp=<timestamp>",
    "pattern": "cdn-asset-mel-1\\\\.airsquare(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/www/site\\\\.css\\\\?timestamp=[\\\\d]+",
  },
  Object {
    "name": "assets.gothamistllc.<domain>/v8/css/sfist.css",
    "pattern": "assets\\\\.gothamistllc(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/v8/css/sfist\\\\.css",
  },
  Object {
    "name": "blog.rudylee.<domain>/stylesheets/screen.css",
    "pattern": "blog\\\\.rudylee(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/stylesheets/screen\\\\.css",
  },
  Object {
    "name": "fathomaway.<domain>/static/cache/css/<hash12>.css",
    "pattern": "fathomaway(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/static/cache/css/[0-9a-f]{12}\\\\.css",
  },
  Object {
    "name": "s.huffpost.<domain>/assets/css.php?f=_quickread.css&v=<version>",
    "pattern": "s\\\\.huffpost(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/assets/css\\\\.php\\\\?f=_quickread\\\\.css&v=([\\\\d.]|[a-f\\\\d])+",
  },
  Object {
    "name": "s.huffpost.<domain>/assets/css.php?f=hptwitter_anywhere.css&9",
    "pattern": "s\\\\.huffpost(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/assets/css\\\\.php\\\\?f=hptwitter_anywhere\\\\.css&9",
  },
  Object {
    "name": "s.huffpost.<domain>/assets/css.php?v=<version>&f=hp_modules%2Fmodule.bubble.css%2Chp_modules%2Fmodule.sponsor-widgets.css%2Chp_modules%2Fgeneral.css%2Chp_modules%2Fmodule.fb_like_button.css%2Cfeatured-posts.css%2Clogos.css%2Ccomments.css%2Cmod-follow.css%2Cfloating_share.css",
    "pattern": "s\\\\.huffpost(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/assets/css\\\\.php\\\\?v=([\\\\d.]|[a-f\\\\d])+&f=hp_modules%2Fmodule\\\\.bubble\\\\.css%2Chp_modules%2Fmodule\\\\.sponsor-widgets\\\\.css%2Chp_modules%2Fgeneral\\\\.css%2Chp_modules%2Fmodule\\\\.fb_like_button\\\\.css%2Cfeatured-posts\\\\.css%2Clogos\\\\.css%2Ccomments\\\\.css%2Cmod-follow\\\\.css%2Cfloating_share\\\\.css",
  },
  Object {
    "name": "s.huffpost.<domain>/assets/css.php?v=<version>&f=verticals%2Fdefault.css%2Ctopnav_new%2Fmodule.edition_nav.css%2Cbasic.css%2Cmasthead.css%2Call_pages.css%2Cmaster.css%2Cbuttons.css%2Chp_modules%2Fmodule.photo_galleries.css%2Cprofiles.css%2Cadd%2Finner.css%2Chp_modules%2Fmodule.threeup.css%2Cemail_share.css%2Ccommercial.css%2Csite_promo.css%2Cpoll_like.css",
    "pattern": "s\\\\.huffpost(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/assets/css\\\\.php\\\\?v=([\\\\d.]|[a-f\\\\d])+&f=verticals%2Fdefault\\\\.css%2Ctopnav_new%2Fmodule\\\\.edition_nav\\\\.css%2Cbasic\\\\.css%2Cmasthead\\\\.css%2Call_pages\\\\.css%2Cmaster\\\\.css%2Cbuttons\\\\.css%2Chp_modules%2Fmodule\\\\.photo_galleries\\\\.css%2Cprofiles\\\\.css%2Cadd%2Finner\\\\.css%2Chp_modules%2Fmodule\\\\.threeup\\\\.css%2Cemail_share\\\\.css%2Ccommercial\\\\.css%2Csite_promo\\\\.css%2Cpoll_like\\\\.css",
  },
  Object {
    "name": "static1.squarespace.<domain>/static/sitecss/<hash>/222/<hash>/<hash>/2755-05142015/<hash>/site.css?&filterFeatures=false",
    "pattern": "static1\\\\.squarespace(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/static/sitecss/[0-9a-f]{10,}/222/[0-9a-f]{10,}/[0-9a-f]{10,}/2755-05142015/[0-9a-f]{10,}/site\\\\.css\\\\?&filterFeatures=false",
  },
  Object {
    "name": "sumome-140a.kxcdn.<domain>/virtual/<hash40>/client/css/sme-popup.css",
    "pattern": "sumome-140a\\\\.kxcdn(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/virtual/[0-9a-f]{40}/client/css/sme-popup\\\\.css",
  },
  Object {
    "name": "sumome-140a.kxcdn.<domain>/virtual/<hash40>/client/css/listbuilder-popup.css",
    "pattern": "sumome-140a\\\\.kxcdn(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/virtual/[0-9a-f]{40}/client/css/listbuilder-popup\\\\.css",
  },
  Object {
    "name": "sumome-140a.kxcdn.<domain>/virtual/<hash40>/client/css/sumome-welcome-popup.css",
    "pattern": "sumome-140a\\\\.kxcdn(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/virtual/[0-9a-f]{40}/client/css/sumome-welcome-popup\\\\.css",
  },
  Object {
    "name": "sumome-140a.kxcdn.<domain>/virtual/<hash40>/client/css/sumome-share-client.css",
    "pattern": "sumome-140a\\\\.kxcdn(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/virtual/[0-9a-f]{40}/client/css/sumome-share-client\\\\.css",
  },
  Object {
    "name": "sumome-140a.kxcdn.<domain>/virtual/<hash40>/client/css/sumome-scrollbox-popup.css",
    "pattern": "sumome-140a\\\\.kxcdn(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/virtual/[0-9a-f]{40}/client/css/sumome-scrollbox-popup\\\\.css",
  },
  Object {
    "name": "www.nomadicmatt.<domain>/wp-content/plugins/searchwp-live-ajax-search/assets/styles/style.css?ver=<version>",
    "pattern": "www\\\\.nomadicmatt(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/wp-content/plugins/searchwp-live-ajax-search/assets/styles/style\\\\.css\\\\?ver=([\\\\d.]|[a-f\\\\d])+",
  },
  Object {
    "name": "www.nomadicmatt.<domain>/wp-content/plugins/woocommerce/assets/css/woocommerce-layout.css?ver=<version>",
    "pattern": "www\\\\.nomadicmatt(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/wp-content/plugins/woocommerce/assets/css/woocommerce-layout\\\\.css\\\\?ver=([\\\\d.]|[a-f\\\\d])+",
  },
  Object {
    "name": "www.nomadicmatt.<domain>/wp-content/themes/nomadicmatt/style.css?ver=<version>",
    "pattern": "www\\\\.nomadicmatt(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/wp-content/themes/nomadicmatt/style\\\\.css\\\\?ver=([\\\\d.]|[a-f\\\\d])+",
  },
  Object {
    "name": "assets-cdn.github.<domain>/assets/github-<hash64>.css",
    "pattern": "assets-cdn\\\\.github(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/assets/github-[0-9a-f]{64}\\\\.css",
  },
  Object {
    "name": "assets.pixolum.<domain>/css/main.min_14a.css",
    "pattern": "assets\\\\.pixolum(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/css/main\\\\.min_14a\\\\.css",
  },
  Object {
    "name": "autodocs.<domain>/assets/application-<hash64>.css",
    "pattern": "autodocs(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/assets/application-[0-9a-f]{64}\\\\.css",
  },
  Object {
    "name": "cdn-asset-mel-1.airsquare.<domain>/www/site.css?timestamp=<timestamp>",
    "pattern": "cdn-asset-mel-1\\\\.airsquare(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/www/site\\\\.css\\\\?timestamp=[\\\\d]+",
  },
  Object {
    "name": "cdn.sstatic.<domain>/Sites/stackoverflow/all.css?v=<version>",
    "pattern": "cdn\\\\.sstatic(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/Sites/stackoverflow/all\\\\.css\\\\?v=([\\\\d.]|[a-f\\\\d])+",
  },
  Object {
    "name": "ucarecdn.<domain>/assets/frontend/public/app.<hash32>.css",
    "pattern": "ucarecdn(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/assets/frontend/public/app\\\\.[0-9a-f]{32}\\\\.css",
  },
  Object {
    "name": "wfaprod-1290.kxcdn.<domain>/themes.css?v=<version>",
    "pattern": "wfaprod-1290\\\\.kxcdn(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/themes\\\\.css\\\\?v=([\\\\d.]|[a-f\\\\d])+",
  },
  Object {
    "name": "www.google.<domain>/uds/api/visualization/1.1/<hash32>/ui+entable+en.css",
    "pattern": "www\\\\.google(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/uds/api/visualization/1\\\\.1/[0-9a-f]{32}/ui\\\\+entable\\\\+en\\\\.css",
  },
  Object {
    "name": "caphunters.<domain>/themes/caphunters/cache/v_92_<hash32>_all.css",
    "pattern": "caphunters(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/themes/caphunters/cache/v_92_[0-9a-f]{32}_all\\\\.css",
  },
  Object {
    "name": "testdomain.<domain>/test.css",
    "pattern": "testdomain(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/test\\\\.css",
  },
  Object {
    "name": "testdomain.<domain>/test.css",
    "pattern": "testdomain(.(ac|co|com|edu|gov|me|net|org))?.[^./]+/test\\\\.css",
  },
]
`;
